File: backend/src/controllers/sermon.controller.ts
import { Request, Response, NextFunction } from 'express';
import prisma from '../prisma';
import { resolveUserId } from '../services/auth';
import { SermonService } from '../services/sermon.service';

export const getSermons = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const book = req.params.book as string;
    let plan = 'FREE';

    try {
      const resolvedUserId = await resolveUserId(req);
      const user = await prisma.user.findUnique({ where: { id: resolvedUserId } });
      if (user) {
        plan = user.plan || 'FREE';
      }
    } catch (authError) {
      console.log('[SermonController] User not authenticated or resolution failed, defaulting to FREE plan');
    }

    const sermons = await SermonService.getSermonsForBook(book);
    
    // Gating logic:
    // FREE tier: only returns top 3 videos
    // BASIC / PREMIUM: returns up to top 6 videos
    const userPlan = plan.toUpperCase();
    const isFree = userPlan === 'FREE';
    const gatedSermons = isFree ? sermons.slice(0, 3) : sermons;

    res.json({
      status: 'success',
      data: {
        sermons: gatedSermons,
        plan: userPlan,
        totalAvailable: sermons.length
      }
    });
  } catch (error) {
    next(error);
  }
};

export const refreshBookSermons = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const book = req.params.book as string;
    const sermons = await SermonService.forceRefresh(book);
    res.json({
      status: 'success',
      data: sermons
    });
  } catch (error) {
    next(error);
  }
};

export const updateUserPlan = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const userId = await resolveUserId(req);
    const { plan } = req.body;
    
    if (!plan || !['FREE', 'BASIC', 'PREMIUM'].includes(plan.toUpperCase())) {
      res.status(400).json({ status: 'error', message: 'Plano inválido. Escolha entre FREE, BASIC ou PREMIUM.' });
      return;
    }

    const updatedUser = await prisma.user.update({
      where: { id: userId },
      data: { plan: plan.toUpperCase() }
    });

    res.json({
      status: 'success',
      data: {
        plan: updatedUser.plan
      }
    });
  } catch (error) {
    next(error);
  }
};

File: src/main/java/com/bibliaviva/domain/bible/SermonController.java
  package com.bibliaviva.domain.bible;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/sermons")
public class SermonController {

    private final SermonService sermonService;

    public SermonController(SermonService sermonService) {
        this.sermonService = sermonService;
    }

    /**
     * Get sermons for a specific Bible book, applying user plan-based gating.
     * FREE user: sees top 3 sermons.
     * BASIC/PREMIUM user: sees top 6 sermons.
     */
    @GetMapping("/{book}")
    public ResponseEntity<Map<String, Object>> getSermons(
            @PathVariable String book,
            @RequestHeader(value = "X-User-Plan", defaultValue = "FREE") String plan) {
        
        List<BookSermon> sermons = sermonService.getSermonsForBook(book);
        
        String userPlan = plan.toUpperCase();
        boolean isFree = "FREE".equals(userPlan);
        
        List<BookSermon> gatedSermons = isFree && sermons.size() > 3 
                ? sermons.subList(0, 3) 
                : sermons;

        Map<String, Object> response = new HashMap<>();
        response.put("status", "success");
        
        Map<String, Object> data = new HashMap<>();
        data.put("sermons", gatedSermons);
        data.put("plan", userPlan);
        data.put("totalAvailable", sermons.size());
        
        response.put("data", data);
        
        return ResponseEntity.ok(response);
    }

    /**
     * Trigger manual background refresh for a book
     */
    @PostMapping("/{book}/refresh")
    public ResponseEntity<Map<String, Object>> refreshSermons(@PathVariable String book) {
        try {
            List<BookSermon> sermons = sermonService.forceRefresh(book);
            Map<String, Object> response = new HashMap<>();
            response.put("status", "success");
            response.put("data", sermons);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, Object> errorResponse = new HashMap<>();
            errorResponse.put("status", "error");
            errorResponse.put("message", e.getMessage());
            return ResponseEntity.internalServerError().body(errorResponse);
        }
    }
}
